http/server.go

```diff
@@ -101,16 +101,30 @@ func (h *handler) start() {
 
 func (h *handler) Watch(ctx context.Context, ready chan bool) {
 RESET:
-	stream := h.client.Watch(context.Background())
+	watchCtx, cncl := context.WithCancel(ctx)
+	stream := h.client.Watch(watchCtx)
 
 	// signal that the watch is ready
 	select {
 	case ready <- true:
 	default:
 	}
 
+	expectedRoundDelayBackoff := time.Minute
+	h.chainInfoLk.RLock()
+	if h.chainInfo != nil {
+		expectedRoundDelayBackoff = h.chainInfo.Period * 2
+	}
+	h.chainInfoLk.RUnlock()
 	for {
-		next, ok := <-stream
+		var next client.Result
+		var ok bool
+		select {
+		case next, ok = <-stream:
+		case <-time.After(expectedRoundDelayBackoff):
+			cncl()
+			goto RESET
```

Comment:
I'd love to get rid of this `goto`. Maybe we could use a helper function? Something like
```
func (h *handler) Watch(ctx context.Context, ready chan bool) {
    for {
        select {
        case <-ctx.Done():
            return
        default:
        }
        h.watchWithTimeout(ctx, ready)
    }
}

func (h *handler) watchWithTimeout(ctx context.Context, ready chan bool) {
    watchCtx, cncl := context.WithCancel(ctx)
    stream := h.client.Watch(watchCtx)
    for {
        var next client.Result
	var ok bool
	select {
        case <-ctx.Done():
            return
	case next, ok = <-stream:
	case <-time.After(expectedRoundDelayBackoff):
             return
	}
	if !ok {
	    h.log.Warn("http_server", "random stream round failed")
	    h.pendingLk.Lock()
            h.latestRound = 0
	    h.pendingLk.Unlock()
	    // backoff on failures a bit to not fall into a tight loop.
	    // TODO: tuning.
	    time.Sleep(watchConnectBackoff)
	    return
         }
         ...
    }
}
```